File: webhooks/utils/query_sqlite_db.py
#!/usr/bin/env python3
"""Print every email stored by the webhook codelab."""

import sqlite3
import sys
from collections.abc import Generator
from contextlib import closing, contextmanager
from pathlib import Path

DATABASE_PATH = Path(__file__).with_name("webhook_codelab_storage.db")


@contextmanager
def _open_read_only_database(
    database_path: Path = DATABASE_PATH,
) -> Generator[sqlite3.Connection]:
    """Helper function that opens a read-only connection to a specified database."""
    if not database_path.is_file():
        raise FileNotFoundError(
            f"Database not found at {database_path}. Run create_sqlite_db.py first."
        )

    # URI mode prevents this query helper and the agent tools that use it from
    # modifying the codelab database.
    database_uri = f"{database_path.resolve().as_uri()}?mode=ro"
    with closing(sqlite3.connect(database_uri, uri=True)) as connection:
        connection.row_factory = sqlite3.Row
        yield connection


@contextmanager
def _open_read_only_database(
    database_path: Path = DATABASE_PATH,
) -> Generator[sqlite3.Connection]:
    """Helper function that opens a read-only connection to a specified database."""
    if not database_path.is_file():
        raise FileNotFoundError(
            f"Database not found at {database_path}. Run create_sqlite_db.py first."
        )

    # URI mode prevents this query helper and the agent tools that use it from
    # modifying the codelab database.
    database_uri = f"{database_path.resolve().as_uri()}?mode=ro"
    with closing(sqlite3.connect(database_uri, uri=True)) as connection:
        connection.row_factory = sqlite3.Row
        yield connection


# Exercise 3: Reuse this read-only query for the recipient-history tool. The
# optional recipient argument limits the result to that recipient's records.
def fetch_emails(
    database_path: Path = DATABASE_PATH,
    recipient: str | None = None,
) -> list[sqlite3.Row]:
    """Return stored emails in insertion order, optionally for one recipient.

    Checking the path first prevents ``sqlite3.connect`` from silently creating
    an empty database when the codelab database has not been initialized.
    """
    query = """
        SELECT *
        FROM emails
        """
    parameters: tuple[str, ...] = ()
    if recipient is not None:
        query += "WHERE recipients = ?\n"
        parameters = (recipient,)
    query += "ORDER BY id"

    with _open_read_only_database(database_path) as connection:
        return connection.execute(query, parameters).fetchall()


def fetch_email_activity(
    database_path: Path = DATABASE_PATH,
) -> list[sqlite3.Row]:
    """Return one email-count row per recipient, ordered by activity."""
    query = """
        SELECT
            recipients AS recipient,
            COUNT(*) AS email_count
        FROM emails
        GROUP BY recipients
        ORDER BY email_count DESC, recipient
        """

    with _open_read_only_database(database_path) as connection:
        return connection.execute(query).fetchall()


def fetch_email_activity(
    database_path: Path = DATABASE_PATH,
) -> list[sqlite3.Row]:
    """Return one email-count row per recipient, ordered by activity."""
    query = """
        SELECT
            recipients AS recipient,
            COUNT(*) AS email_count
        FROM emails
        GROUP BY recipients
        ORDER BY email_count DESC, recipient
        """

    with _open_read_only_database(database_path) as connection:
        return connection.execute(query).fetchall()


# Exercise 3: Add a read-only query helper here that groups stored emails by
# recipient and returns each recipient with its email count for the summary tool.
def print_emails(emails: list[sqlite3.Row]) -> None:
    """Print email rows in an aligned table with database field headings."""
    field_names = tuple(emails[0].keys())
    rows = [tuple(str(email[field]) for field in field_names) for email in emails]
    column_widths = [
        max(len(field), *(len(row[index]) for row in rows))
        for index, field in enumerate(field_names)
    ]

    def format_row(values: tuple[str, ...]) -> str:
        return " | ".join(
            value.ljust(column_widths[index]) for index, value in enumerate(values)
        )

    print(format_row(field_names))
    print("-+-".join("-" * width for width in column_widths))
    for row in rows:
        print(format_row(row))


def main() -> int:
    """Query the codelab database and print readable email records."""
    try:
        emails = fetch_emails()
    except (FileNotFoundError, sqlite3.Error) as error:
        print(f"Unable to query stored emails: {error}", file=sys.stderr)
        return 1

    if not emails:
        print(f"No emails found in {DATABASE_PATH}")
    else:
        print_emails(emails)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: webhooks/java/src/main/java/io/orkes/codelab/webhooks/utils/QuerySqliteDb.java
package io.orkes.codelab.webhooks.utils;

import org.sqlite.SQLiteConfig;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

/** Prints emails stored by the DB for the webhook codelab. */
public final class QuerySqliteDb {

    /** The local database file, created by the create-db step in this project's folder. */
    public static final Path DATABASE_PATH = Path.of("webhook_codelab_storage.db").toAbsolutePath();

    private QuerySqliteDb() {}

    /** Return every stored email in insertion order. */
    public static List<Map<String, Object>> fetchEmails() throws IOException, SQLException {
        return fetchEmails(null);
    }

    /**
     * Return stored emails in insertion order, for one recipient unless {@code recipient} is null.
     */
    public static List<Map<String, Object>> fetchEmails(String recipient)
            throws IOException, SQLException {
        String query =
                """
                SELECT *
                FROM emails
                """
                        + (recipient != null ? "WHERE recipients = ?\n" : "")
                        + "ORDER BY id";

        try (Connection connection = openReadOnlyDatabase();
                var statement = connection.prepareStatement(query)) {
            if (recipient != null) {
                statement.setString(1, recipient);
            }
            try (ResultSet resultSet = statement.executeQuery()) {
                return readRows(resultSet);
            }
        }
    }

    /** Return one email-count row per recipient, ordered by activity. */
    public static List<Map<String, Object>> fetchEmailActivity() throws IOException, SQLException {
        String query =
                """
                SELECT
                    recipients AS recipient,
                    COUNT(*) AS email_count
                FROM emails
                GROUP BY recipients
                ORDER BY email_count DESC, recipient
                """;

        try (Connection connection = openReadOnlyDatabase();
                var statement = connection.prepareStatement(query);
                ResultSet resultSet = statement.executeQuery()) {
            return readRows(resultSet);
        }
    }

    /** Print email rows in an aligned table with database field headings. */
    public static void printEmails(List<Map<String, Object>> emails) {
        List<String> fieldNames = List.copyOf(emails.getFirst().keySet());
        List<List<String>> rows =
                emails.stream()
                        .map(
                                email ->
                                        fieldNames.stream()
                                                .map(field -> String.valueOf(email.get(field)))
                                                .toList())
                        .toList();
        // Each column is as wide as its longest value, counting the heading as a value.
        int[] columnWidths =
                IntStream.range(0, fieldNames.size())
                        .map(
                                index ->
                                        Stream.concat(Stream.of(fieldNames), rows.stream())
                                                .mapToInt(row -> row.get(index).length())
                                                .max()
                                                .orElseThrow())
                        .toArray();

        System.out.println(formatRow(fieldNames, columnWidths));
        System.out.println(
                Arrays.stream(columnWidths)
                        .mapToObj("-"::repeat)
                        .collect(Collectors.joining("-+-")));
        rows.forEach(row -> System.out.println(formatRow(row, columnWidths)));
    }

    /** Query the codelab database and print readable email records. */
    public static int run() {
        List<Map<String, Object>> emails;
        try {
            emails = fetchEmails();
        } catch (IOException | SQLException error) {
            System.err.println("Unable to query stored emails: " + error.getMessage());
            return 1;
        }

        if (emails.isEmpty()) {
            System.out.println("No emails found in " + DATABASE_PATH);
        } else {
            printEmails(emails);
        }
        return 0;
    }

    /**
     * Open a read-only connection to the codelab database. Checking the path first gives a clear
     * error when the database has not been created yet.
     */
    private static Connection openReadOnlyDatabase() throws IOException, SQLException {
        if (!Files.isRegularFile(DATABASE_PATH)) {
            throw new FileNotFoundException(
                    "Database not found at " + DATABASE_PATH + ". Run the create-db step first.");
        }

        // Read-only mode prevents this query helper and the agent tools that use it from
        // modifying the codelab database.
        SQLiteConfig config = new SQLiteConfig();
        config.setReadOnly(true);
        return DriverManager.getConnection("jdbc:sqlite:" + DATABASE_PATH, config.toProperties());
    }

    /** Read every result row as a map keyed by column name, in column order. */
    private static List<Map<String, Object>> readRows(ResultSet resultSet) throws SQLException {
        // Plain loops read the ResultSet: it has no stream API, and its getters throw the checked
        // SQLException, which a stream's lambdas cannot throw.
        var metadata = resultSet.getMetaData();
        int columnCount = metadata.getColumnCount();
        List<Map<String, Object>> rows = new ArrayList<>();
        while (resultSet.next()) {
            Map<String, Object> row = new LinkedHashMap<>();
            for (int column = 1; column <= columnCount; column++) {
                row.put(metadata.getColumnLabel(column), resultSet.getObject(column));
            }
            rows.add(row);
        }
        return rows;
    }

    /** Join one row's values with column separators, padding each value to its column's width. */
    private static String formatRow(List<String> values, int[] columnWidths) {
        return IntStream.range(0, values.size())
                .mapToObj(
                        index -> String.format("%-" + columnWidths[index] + "s", values.get(index)))
                .collect(Collectors.joining(" | "));
    }
}
